### DISCUSSION TITLE
Optimizer stops after one Objective Function call

### DISCUSSION THREAD

**Discussion by AbhijithMoni** [mdolab/pyoptsparse#318] [Q&A]
I'm trying to optimize an airfoil with set of different airfoil co-ordinate points that has (64, 236) features using pyoptsparse within a given bound. The objective is to reduce drag and obtain the corresponding co-ordinate points.  But the optimizer stops after one obj function call. Can anyone please tell me what am I doing wrong here. I have attached the code below:

```
def objfunc(xdict):

    xvar = []

    for i in range(236):

        xvar.append(xdict['x{}'.format(i)])

    xvar=np.array(xvar)

    xvar= xvar.reshape(1, -1)

    

    cd_model = pickle.load(open('gpr_cd.sav', 'rb'))
    
    thickness_model = pickle.load(open('gpr_thickness.sav', 'rb'))


    funcs = {}

    funcs["obj"] = cd_model.predict(xvar)

  
    funcs["con1"] =  thickness_model.predict(xvar)

    fail = False 

    return funcs, fail

optProb = Optimization('Airfoil opti',objfunc)

sample = open('limits.dat', "r")
temp=sample.readlines()
sample.close()
limits = []
for line in temp:
    ln=line.split()
    limits.append([float(i) for i in ln])
limits=np.array(limits)

lwlim=limits[:,0]
uplim=limits[:,1]




for i in range(236):
    optProb.addVar("x{}".format(i),"c", value=0, lower=lwlim[i], upper=uplim[i])


optProb.addObj("obj")


optProb.addCon("con1", lower=0.12 , upper=None)   #con1: maximum thickness

print('Running optimization...')

parser = argparse.ArgumentParser()
parser.add_argument("--opt", help="optimizer", type=str, default="SLSQP")
parser.add_argument("--storeHistory", help="option to store history", action="store_true")
args = parser.parse_args()
optOptions = {"ACC": 1e-06,"IPRINT": -1}


opt = OPT(args.opt, options=optOptions)


sol = opt(optProb, sens="FD", storeHistory='hisoty_airfoil.hst')

print(sol)
```

```


    Solution: 
--------------------------------------------------------------------------------
    Total Time:                    2.4678
       User Objective Time :       0.0327
       User Sensitivity Time :     1.4744
       Interface Time :            0.4668
       Opt Solver Time:            0.4939
    Calls to Objective Function :       1
    Calls to Sens Function :            1
```

Any help is appreciated. 

Many thanks 
Abhijith

**Comment by marcomangano** [comment]
Hi Abhijith,
it is hard to tell what is wrong from here, but my guess is that something is wrong with the gradient, the optimizer gets a vector of zeros for the sensitivities and completes the optimization at the first iteration.
Some tips for debugging:
- check all the output info from SLSQP
- print the values of the DVs and the evaluated functions within `objFunc` to make sure the set up is correct there. As you are using the "automatic" finite-difference sensitivities from pyOptSparse, the code will call repeatedly your `objFunc` to compute the gradient.
- You can check the sensitivities from the history file with something like:
```
    from sqlitedict import SqliteDict
    from collections import OrderedDict

    db = SqliteDict(<path-to-history-file>)
    hist_db = copy.deepcopy(OrderedDict(db))
    db.close()
    print(hist_db["1"])
```
see the [documentation](https://mdolab-pyoptsparse.readthedocs-hosted.com/en/latest/api/history.html) to understand the structure of the history file. "1" here should be your first gradient evaluation, and your sensitivities are stored as nested dictionaries.

Good luck,
               Marco

**Reply by AbhijithMoni** [reply]
Thanks for your reply @marcomangano. I will try the method as suggested to fix the code. 

I really appreciate your time and help. 

Thanks 
Abhijith
